Fix to lean 4 documentation for calculational proofs (#391)
leanprover-community/leanprover-community.github.io: templates/extras/calc.md

# How to use calc

`calc` is an environment -- so a "mode" like tactic mode, term mode and
[conv mode](conv.html). Documentation and basic examples for how to use
it are in Theorem Proving In Lean, in
[section 4](https://lean-lang.org/theorem_proving_in_lean4/quantifiers_and_equality.html#calculational-proofs).

Basic example usage:

```lean
example (a b c : ℕ) (H1 : a = b + 1) (H2 : b = c) : a = c + 1 :=
calc a = b + 1 := H1

>      _ = c + 1 := by rw[H2]

Reviewer: This style is preferred:
```suggestion
     _ = c + 1 := by rw [H2]
```

> calc a = b + 1 := H1
>      _ = c + 1 := by rw[H2]

Reviewer: Can you adapt to putting the `calc` on its own line here and everywhere else? This is preferred style now
```suggestion
calc
  a = b + 1 := H1
  _ = c + 1 := by rw[H2]
```

```

> `calc` is also available in tactic mode.

Reviewer: ```suggestion
`calc` is also available in tactic mode. You can leave `?_`s to create a
```

new goal:
```lean

> example (a b c : ℕ) (H1 : a = b + 1) (H2 : b = c) : a = c + 1 := by
>   calc a = b + 1 := H1
>        _ = c + 1 := by rw[H2]

Reviewer: ```suggestion
example (a b c : ℕ) (H1 : a = b + 1) (H2 : b = c) : a = c + 1 := by
  calc
    a = b + 1 := ?_
    _ = c + 1 := ?_
  · exact H1
  · rw [H2]
```

for missing justifications. `sorry` will supress error messages
entirely, while `_` will generate a guiding error message.

If the structure of calc is incorrect (e.g., missing `:=` or the
justification after it), you may see error messages that are obscure
and/or red squiggles that end up under a random `_`. To avoid these,
you might first populate a skeleton proof such as:

```lean
example (A B C D : ℝ ) : A = D :=
calc A = B := sorry
     _ = C := _
     _ = D := sorry
```

and then fill in the `sorry` and `_` gradually.

> (Idle thought: could one write a VS Code snippet to write this skeleton?)

Reviewer 2: This comment is outdated. There is now a code action doing it. More generally this documentation should now explain the calc widget.

Author: Do you have details of the calc widget? I can't work it out

Author: Didn't know about this having a look into this

Author: Sorry can't work out how to use this

Author: Apparently you've made a video about this I'll go find it

Author: added a link to a video of you demonstrating this.

for missing justifications. `sorry` will supress error messages
entirely, while `_` will generate a guiding error message.

If the structure of calc is incorrect (e.g., missing `:` or the
If the structure of calc is incorrect (e.g., missing `:=` or the
justification after it), you may see error messages that are obscure
and/or red squiggles that end up under a random `...`. To avoid these,
and/or red squiggles that end up under a random `_`. To avoid these,
you might first populate a skeleton proof such as:

```lean
example (A B C D : ℝ ) : A = D :=
calc A = B : sorry
...    = C : _
...    = D : sorry
calc A = B := sorry
     _ = C := _
     _ = D := sorry
```

and then fill in the `sorry` and `_` gradually.

In tactic mode calc should be terminated with a comma:
```lean
have H : A = D,
{ calc A = B : sorry
  ...    = C : sorry
  ...    = D : _,
  sorry },
```

> and the `_` can be left in as they generate a subgoal to be resolved

Reviewer 3: This sentence is still true (or at least, it's true with `?_` instead of `_`.

Author: Thanks will try this out soon.

Author: Okay added that information back in

for missing justifications. `sorry` will supress error messages
entirely, while `_` will generate a guiding error message.

If the structure of calc is incorrect (e.g., missing `:=` or the
justification after it), you may see error messages that are obscure
and/or red squiggles that end up under a random `_`. To avoid these,

> you might first populate a skeleton proof such as:

Reviewer: I'm not actually sure this is true anymore. You should test precisely what happens in several common scenarios (missing `_` after the `:=`, missing `:=`, missing `_` before the relation...) and report back.

```lean
example (A B C D : ℝ ) : A = D :=
calc A = B := sorry
     _ = C := _
     _ = D := sorry
```

and then fill in the `sorry` and `_` gradually.

> In tactic mode a `?_` can be used to generate a subgoal to be resolved.

Reviewer: That's already mentioned above

> In tactic mode a `?_` can be used to generate a subgoal to be resolved.

Reviewer 2: ```suggestion
```

```lean
example (A B C D : ℝ ) : A = D := by
     calc A = B := sorry
          _ = C := ?_
          _ = D := sorry
     sorry
```

> There is also a Calc widget that makes this easier here is a [video](https://youtu.be/8MFGhOWeCNE?t=1834). demonstrating it.

Reviewer: ```suggestion
There is also a Calc widget that makes this easier here is a [video](https://youtu.be/8MFGhOWeCNE?t=1834) demonstrating it.
```

## Using operators other than equality

Many of the basic examples in TPIL use equality for most or all of
the operators, but actually `calc` will work with any relation for which
we have created an instance of `Trans`

```lean
def r : ℕ → ℕ → Prop := sorry
variable (a b c: ℕ)
def r_trans {a b c} (h₁ : r a b) (h₂ : r b c) : r a c := sorry

instance : Trans r r r where
  trans := r_trans

infix:50 "***" => r

example (a b c : ℕ) (H1 : a *** b) (H2 : b *** c) : a *** c :=
calc a *** b := H1
     _ *** c := H2
```

## Using more than one operator

This is possible, for example:

```lean
theorem T2 (a b c d : ℕ)
  (h1 : a = b) (h2 : b ≤ c) (h3 : c + 1 < d) : a < d := 
  calc
    a = b     := h1
    _ < b + 1 := Nat.lt_succ_self b
    _ ≤ c + 1 := Nat.succ_le_succ h2
    _ < d     := h3

 ```

What is actually going on here? The proofs themselves are not a mystery,
for example `Nat.succ_le_succ h2` is a proof of `b + 1  ≤ c + 1`. The
clever part is that lean can put all of these together to correctly
deduce that if `U = V < W ≤ X < Y` then `U < Y`. The way this is done,

> Kevin thinks (can someone verify this?) is that Lean continually tries

Reviewer: Can you verify this?

> the corresponding transitivity statement is tagged `[trans]`:

Reviewer 2: This is outdated. This attribute no longer exists.

Author: Whoops sorry missed that one

is `=`. Lean knows

```lean
#check trans_rel_right -- {α : Sort u} {a b c : α} (r : α → α → Prop) (h₁ : a = b) (h₂ : r b c) : r a c
#check trans_rel_left -- {α : Sort u} {a b c : α} (r : α → α → Prop) (h₁ : r a b) (h₂ : b = c) : r a c

and (Kevin believes) uses them if one of the operators is an equality operator. If however neither
operator is the equality operator, Lean looks through the instances of
`Trans` and applies these instead.

## Using user-defined operators

> It is as simple as tagging the relevant adding new `Trans`. For example

Reviewer: This sentence is grammatically incorrect

> It is as simple as tagging the relevant adding new `Trans`. For example

Reviewer 2: ```suggestion
It is as simple as creating the relevant `Trans` instances. For example
```
